File: extract_audio_from_mp4/extractAudioFromMP4.go
package main

import (
    "bufio"
    "bytes"
    "encoding/binary"
    "fmt"
    "log"
    "os"
    "os/exec"
    "path/filepath"
    "strings"

    "github.com/go-audio/audio"
    "github.com/go-audio/wav"
)

// extractAudioFromMP4 从 MP4 文件中提取音频
func extractAudioFromMP4(mp4Path string, tempAudioPath string) error {
    cmd := exec.Command("ffmpeg", "-i", mp4Path, "-vn", "-acodec", "pcm_s16le", "-ar", "44100", "-ac", "1", tempAudioPath)
    var out bytes.Buffer
    cmd.Stdout = &out
    cmd.Stderr = &out
    err := cmd.Run()
    if err != nil {
        log.Printf("提取音频时出错: %v, 输出: %s", err, out.String())
        return err
    }
    return nil
}

// readAudioFile 读取音频文件
func readAudioFile(audioPath string) ([]int16, int, error) {
    file, err := os.Open(audioPath)
    if err != nil {
        return nil, 0, err
    }
    defer file.Close()

    decoder := wav.NewDecoder(bufio.NewReader(file))
    if !decoder.IsValidFile() {
        return nil, 0, fmt.Errorf("不是有效的 WAV 文件: %s", audioPath)
    }

    if err := decoder.FullDecode(); err != nil {
        return nil, 0, err
    }

    samples := make([]int16, len(decoder.Data.Floats()))
    for i, f := range decoder.Data.Floats() {
        samples[i] = int16(f * 32767)
    }

    return samples, int(decoder.SampleRate), nil
}

// correlate 互相关函数
func correlate(signal1, signal2 []int16) []float64 {
    n := len(signal1)
    m := len(signal2)
    result := make([]float64, n-m+1)

    for i := 0; i <= n-m; i++ {
        sum := 0.0
        for j := 0; j < m; j++ {
            sum += float64(signal1[i+j]) * float64(signal2[j])
        }
        result[i] = sum
    }

    return result
}

// findAudioSegments 在 MP4 音频中查找 MP3 音频片段的位置
func findAudioSegments(mp4Audio []int16, mp4SR int, mp3Audio []int16, mp3SR int, threshold float64) [][2]float64 {
    if mp4SR != mp3SR {
        log.Println("采样率不一致，暂不支持重采样")
        return nil
    }

    corr := correlate(mp4Audio, mp3Audio)
    maxCorr := 0.0
    for _, c := range corr {
        if c > maxCorr {
            maxCorr = c
        }
    }

    segments := [][2]float64{}
    for i, c := range corr {
        if c >= maxCorr*threshold {
            startTime := float64(i) / float64(mp4SR)
            endTime := startTime + float64(len(mp3Audio))/float64(mp4SR)
            segments = append(segments, [2]float64{startTime, endTime})
        }
    }

    return segments
}

// findSimilarSegments 查找文件夹中所有 MP4 文件中包含 MP3 音频片段的部分
func findSimilarSegments(folderPath string, mp3Path string) {
    mp3Audio, mp3SR, err := readAudioFile(mp3Path)
    if err != nil {
        log.Printf("读取 MP3 文件时出错: %v", err)
        return
    }

    err = filepath.Walk(folderPath, func(path string, info os.FileInfo, err error) error {
        if err != nil {
            return err
        }
        if !info.IsDir() && strings.HasSuffix(strings.ToLower(info.Name()), ".mp4") {
            tempAudioPath := "temp_audio.wav"
            if err := extractAudioFromMP4(path, tempAudioPath); err != nil {
                return nil
            }

            mp4Audio, mp4SR, err := readAudioFile(tempAudioPath)
            if err != nil {
                log.Printf("读取 MP4 音频时出错: %v", err)
                os.Remove(tempAudioPath)
                return nil
            }

            segments := findAudioSegments(mp4Audio, mp4SR, mp3Audio, mp3SR, 0.8)
            if len(segments) > 0 {
                fmt.Printf("文件名: %s\n", info.Name())
                fmt.Printf("出现次数: %d\n", len(segments))
                for i, segment := range segments {
                    fmt.Printf("第 %d 次出现时间区间: %.2fs - %.2fs\n", i+1, segment[0], segment[1])
                }
            } else {
                fmt.Printf("文件名: %s, 未找到匹配的音频片段。\n", info.Name())
            }

            os.Remove(tempAudioPath)
        }
        return nil
    })

    if err != nil {
        log.Printf("遍历文件夹时出错: %v", err)
    }
}

func main() {
    folderPath := "./"   // MP4 所在的文件夹
    mp3Path := "123.MP3" // 提供的 MP3 片段采样
    findSimilarSegments(folderPath, mp3Path)
}

File: extract_audio_from_mp4/extract_audio_from_mp4.py
import os
import librosa
import numpy as np
from moviepy.editor import VideoFileClip
from scipy.signal import correlate
import contextlib

"""
@此文件目的是：查找MP4中是否含有 提供的MP3音频片段  匹配几次，起止时间
"""
def extract_audio_from_mp4(mp4_path):
    """
    从 MP4 文件中提取音频
    :param mp4_path: MP4 文件路径
    :return: 音频信号和采样率
    """
    clip = VideoFileClip(mp4_path)
    audio = clip.audio
    with open(os.devnull, 'w') as fnull, contextlib.redirect_stdout(fnull), contextlib.redirect_stderr(fnull):
        audio.write_audiofile("temp_audio.wav")
    y, sr = librosa.load("temp_audio.wav")
    os.remove("temp_audio.wav")
    return y, sr

def find_audio_segments(mp4_audio, mp4_sr, mp3_audio, mp3_sr, downsample_rate=4, threshold=0.8):
    """
    在 MP4 音频中查找所有 MP3 音频片段的位置
    :param mp4_audio: MP4 音频信号
    :param mp4_sr: MP4 音频采样率
    :param mp3_audio: MP3 音频信号
    :param mp3_sr: MP3 音频采样率
    :param downsample_rate: 下采样率
    :param threshold: 匹配阈值
    :return: 匹配的起始和结束时间列表（秒）
    """
    # 调整 MP3 音频采样率与 MP4 音频一致
    mp3_audio_resampled = librosa.resample(mp3_audio, orig_sr=mp3_sr, target_sr=mp4_sr)

    # 下采样以减少计算量
    mp4_audio_downsampled = mp4_audio[::downsample_rate]
    mp3_audio_downsampled = mp3_audio_resampled[::downsample_rate]

    # 使用互相关进行匹配
    correlation = correlate(mp4_audio_downsampled, mp3_audio_downsampled, mode='valid')
    max_corr = np.max(correlation)
    segments = []
    start_index = 0
    while True:
        current_correlation = correlation[start_index:]
        if len(current_correlation) == 0:
            break
        max_index = np.argmax(current_correlation) + start_index
        if correlation[max_index] < max_corr * threshold:
            break

        # 计算实际的起始和结束索引
        actual_start_index = max_index * downsample_rate
        actual_end_index = actual_start_index + len(mp3_audio_resampled)

        # 转换为时间（秒）
        start_time = actual_start_index / mp4_sr
        end_time = actual_end_index / mp4_sr
        segments.append((start_time, end_time))

        # 跳过已经匹配的部分继续查找
        start_index = max_index + len(mp3_audio_downsampled)

    return segments

def find_similar_segments(folder_path, mp3_path):
    """
    查找文件夹中所有 MP4 文件中包含 MP3 音频片段的部分
    :param folder_path: 文件夹路径
    :param mp3_path: MP3 文件路径
    """
    mp3_audio, mp3_sr = librosa.load(mp3_path)
    for root, dirs, files in os.walk(folder_path):
        for file in files:
            if file.endswith('.mp4'):
                mp4_path = os.path.join(root, file)
                mp4_audio, mp4_sr = extract_audio_from_mp4(mp4_path)
                segments = find_audio_segments(mp4_audio, mp4_sr, mp3_audio, mp3_sr)
                if segments:
                    print(f"文件名: {file}")
                    print(f"出现次数: {len(segments)}")
                    for i, (start_time, end_time) in enumerate(segments, start=1):
                        print(f"第 {i} 次出现时间区间: {start_time:.2f}s - {end_time:.2f}s")
                else:
                    print(f"文件名: {file}, 未找到匹配的音频片段。")

# 示例使用
folder_path = './'
mp3_path = '123.MP3'
find_similar_segments(folder_path, mp3_path)
